module/VuFind/src/VuFind/Auth/AbstractShibboleth.php

```diff
@@ -0,0 +1,362 @@
+    {
+        // Throw an exception if the required username setting is missing.
+        $shib = $this->config->Shibboleth;
+        if (!isset($shib->username) || empty($shib->username)) {
+            throw new AuthException(
+                "Shibboleth username is missing in your configuration file."
+            );
+        }
+
+        // Throw an exception if no login endpoint is available.
+        if (!isset($shib->login)) {
+            throw new AuthException(
+                'Shibboleth login configuration parameter is not set.'
+            );
+        }
+    }
+
+    /**
+     * Attempt to authenticate the current user.  Throws exception if login fails.
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @throws AuthException
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
+     */
+    public function authenticate($request)
+    {
+        // Check if username is set.
+        $shib = $this->getShibbolethConfiguration($request);
+        $username = $this->getAttribute($request, $shib['username']);
+        if (empty($username)) {
+            $this->debug(
+                "No username attribute ({$shib['username']}) present in request: "
+                . print_r($request->getServer()->toArray(), true)
+            );
+            throw new AuthException('authentication_error_admin');
+        }
+
+        // Check if required attributes match up:
+        foreach ($this->getRequiredAttributes($shib) as $key => $value) {
+            if (!preg_match(
+                '/' . $value . '/',
+                $this->getAttribute($request, $key)
+            )
+            ) {
+                $this->debug(
+                    "Attribute '$key' does not match required value '$value' in"
+                    . ' request: ' . print_r($request->getServer()->toArray(), true)
+                );
+                throw new AuthException('authentication_error_denied');
+            }
+        }
+
+        // If we made it this far, we should log in the user!
+        $user = $this->getUserTable()->getByUsername($username);
+
+        // Variable to hold catalog password (handled separately from other
+        // attributes since we need to use saveCredentials method to store it):
+        $catPassword = null;
+
+        // Has the user configured attributes to use for populating the user table?
+        foreach ($this->attribsToCheck as $attribute) {
+            if (isset($shib[$attribute])) {
+                $value = $this->getAttribute($request, $shib[$attribute]);
+                if ($attribute == 'email') {
+                    $user->updateEmail($value);
+                } elseif ($attribute == 'cat_username' && isset($shib['prefix'])) {
+                    $user->cat_username = $shib['prefix'] . '.'
+                        . (($value === null) ? '' : $value);
+                } elseif ($attribute != 'cat_password') {
```

Comment:
I'd switch this around and check for `} elseif ($attribute === 'cat_password') {` here and handle other values in the final else block. I know it's just a copy from old code, but it just bugs me. Feel free to ignore if you'd rather keep the moved code intact. :)